File: solution3.py
# Time Complexity : O(log(n))
# Space Complexity : O(1)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach

class Solution:
    def find_peak_element(self, nums):
        
        left,right=0,len(nums)-1
        
        # binary search
        while left < right:
            mid = left + (right-left)//2
            # always going towards increasing trend, a bit counter-intuitive but we always get a peak value
            if nums[mid] > nums[mid+1]:
                right=mid
            else:
                left=mid+1
        return left

File: solution3.java
class Solution {
    public int findPeakElement(int[] nums) {
        int low=0;
        int high = nums.length-1;
        return binarySearch(nums, low, high);
    }
    public int binarySearch(int[] nums, int low, int high){
        int mid;
        while(low<=high){
        mid = low+(high-low)/2;
        if((mid==high||nums[mid]>nums[mid+1]) && (mid==low || nums[mid]>nums[mid-1])){
            return mid;
        }
        else if(nums[mid]<nums[mid+1]){
            low=mid+1;
        }
        else{
            high=mid-1;
        }
        }
        return 28301;
    }
}
